File: leetcode/python/547_number_of_provinces.py
class Solution:
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        visited = set[int]()
        count: int = 0
        for i in range(len(isConnected)):
            if i in visited:
                continue
            
            count += 1
            q: list[int] = [i]
            while q:
                p = q.pop()
                if p in visited:
                    continue
                visited.add(p)
                for j, link in enumerate(isConnected[p]):
                    if link == 1:
                        q.append(j)
        return count

if __name__ == '__main__':
    s = Solution()
    print(s.findCircleNum([[1,1,0],[1,1,0],[0,0,1]]))
    print(s.findCircleNum([[1,0,0],[0,1,0],[0,0,1]]))

File: leetcode/rust/src/547_number_of_provinces.rs
pub fn find_circle_num(is_connected: Vec<Vec<i32>>) -> i32 {
    let mut visited = vec![false; is_connected.len()];
    let mut count = 0;
    for i in 0..is_connected.len() {
        if !visited[i] {
            dfs(&is_connected, &mut visited, i);
            count += 1;
        }
    }
    count
}

pub fn dfs(is_connected: &Vec<Vec<i32>>, visited: &mut Vec<bool>, i: usize) {
    for j in 0..is_connected.len() {
        if is_connected[i][j] == 1 && !visited[j] {
            visited[j] = true;
            dfs(is_connected, visited, j);
        }
    }
}
